llvm/test/Transforms/InstCombine/printf-addrspace.ll

```diff
@@ -0,0 +1,35 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; Test that the printf library call simplifier works correctly.
+;
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:32:64-f32:32:32-f64:32:64-v64:64:64-v128:128:128-a0:0:64-f80:128:128-G200"
+
+@hello_world = addrspace(200) constant [13 x i8] c"hello world\0A\00"
+@percent_s = addrspace(200) constant [4 x i8] c"%s\0A\00"
+
+declare i32 @printf(ptr addrspace(200) , ...)
+declare i32 @puts(ptr addrspace(200))
```

Comment:
I'm concerned this isn't working the way we want.

`puts` takes a C `char*`, and the C library's `char*` is in some specific address-space.  TargetLibraryInfo should inherently know the correct address-space for puts... and on usual targets, the inherent address-space should be zero.  So TargetLibraryInfo should refuse to recognize this declaration, and the call shouldn't be transformed.  The concept here is similar to `size_t`: the same way there's only one `size_t` on any given target, there's also only one `char*`.

Similarly, we shouldn't need to pass a type into emitMalloc and friends: malloc returns a `void*`, and a C `void*` only has one possible associated address-space.

Reply:
CHERI purecap targets don't use address space 0 as the default address space for C pointers. (This may seem like an unusual choice, but there are reasons for it that are somewhat beyond the scope of one PR). My goal is to make sure this works. Ensuring propagation of address spaces is one way of achieving that: it essentially puts the burden of constructing the IR correctly on the frontend, and the middle-end will simply maintain it.

 An alternative approach would be to add something like `getVoidPtrTy` to `TargetLibraryInfo` and have everything use it. This may be net slightly less code, but also means that there are now two sources of truth.

Reply:
The problem with *not* having `TargetLibraryInfo::getVoidPtrTy()` is that you accidentally inherit information from other places.  If the source of that inheritance is other C library functions, you're fine, I guess, but that's extremely hard to audit.  For example, you call emitPutS() with a pointer with address-space getDefaultGlobalsAddressSpace().  Is that right?  We don't know, and nothing will ever check it.

I don't think having a separate source of truth for getVoidPtrTy() is really a big deal compared to all the other assumptions that TargetLibraryInfo makes.  We can make clang codegen check that the TargetLibraryInfo is correct if you're concerned it'll fall out of sync somehow.

Reply:
I took a quick look into making this change, and it breaks a few in-tree tests (which all look like the originated from CHERI targets...). What I'd like to do is, in a separate change, add the target triple bits needed to make it possible to detect CHERI from the target triple, then adapt the tests to use that properly, then make the switch to `getVoidPtrType()` with an override when CHERI is indicated by the triple.

Reply:
See https://github.com/llvm/llvm-project/pull/118927

Reply:
In the past I've been wondering if the cleanest approach here would be to have a "libcalls" address space in the data layout. This would also apply to the result of malloc and free arg. Might be useful for GPU targets if they don't use zero.

Reply:
This also shows up in atomic libcall handling. I don't really care whether it's called the "libcalls address space", "C ABI address space", "default address space" or some other thing like that, but it's a long-standing issue that LLVM only has the fine-grained definition-based address spaces (alloca, global, program) but no use-based / generic address spaces beyond 0, which CHERI backends do not use for various technical reasons (whether that's the right choice is a different matter, but this should be possible to do). I've always felt there should be some sort of `-D<address space>` (if using "default" above) which is the address space of a pointer without any further context, and is what you can then map C's `T *` to. A `-L<address space>` for libcalls might be fine, though I'd be concerned it's still overly narrow.

@resistor Please tag @arichardson and myself on future CHERI-related PRs. Between us we have a lot of experience with CHERI LLVM and I want to ensure you're making upstream changes we agree with rather than going behind our backs. It was a surprise to me at least that these PRs had been filed and merged.

Reply:
See https://github.com/llvm/llvm-project/pull/88182 for some related discussion, in particular https://github.com/llvm/llvm-project/pull/88182#issuecomment-2045975125 onwards. It's important for anyone trying to upstream CHERI-related things to be aware of past discussion on this topic, and unless you want to search the internet for everything related to CHERI you only get that by talking to the people who have been involved in those kinds of discussions.

Reply:
> So TargetLibraryInfo should refuse to recognize this declaration, and the call shouldn't be transformed. 

Well, that's not what came out of https://reviews.llvm.org/D95142, but that was in part due to not having any way to get the right address space for a libcall. What *was* universally agreed though was that picking a random one of the fine-grained address spaces was not correct (the initial version of the patch even used the globals address space just like this one here).

Reply:
> that was in part due to not having any way to get the right address space for a libcall

Right... like I mentioned before, I think we do want to tie libcalls to a specific address-space.  It's way too easy to make mistakes otherwise in an environment with multiple address-spaces.

> What was universally agreed though was that picking a random one of the fine-grained address spaces was not correct (the initial version of the patch even used the globals address space just like this one here).

Yes, none of the existing address-spaces in the datalayout matches the concept we want here.